docs/branching.md

```diff
@@ -0,0 +1,32 @@
+### Branching and deployments model
+
+This document aims to discuss and document the relations between branches and deployments of `elixir-omg`, with respect to branches and deployments of `plasma-contracts`.
+
+It is a refinement of the [OIP4 branching model](https://github.com/omisego/OIP/blob/master/0004-ewallet-release-and-versioning.md), applicable to `elixir-omg` and `plasma-contracts` versioning.
+
+This takes `v0.1` and `v0.2` as an example of two versions, where one is deployed in `staging`/`Ari` and the other is the upcoming upgrade and is deployed in `devel`.
+
+Rationale:
+- the history and the relations between the versions are readable and accountable,
+- we can predictably sync our respective watchers/run child chain servers against deployed contracts
+- and at the same time be able to move with `master`s quickly
+
+("->" means that the `mix.exs` holds the following branch)
+
+- `elixir-omg/v0.1` -> `plasma-contracts/v0.1`, that's understood, the `mix.lock` points to a specific frozen version, no changes expected there
+- `elixir-omg/master` -> `plasma-contracts/master`, mix lock points to the contracts ~deployed on `devel`~ we want to CI-check with `elixir-omg`. So it might be newer than what contract is deployed on `devel`
+- `elixir-omg/v0.2` -> `plasma-contracts/v0.2`, the commit that fixes *the branch `plasma-contracts/v0.2` in `elixir-omg`* will need to be constantly rebased onto `master`, as long we keep looping features into `elixir-omg/v0.2`. *The mix lock will always point to the contract commit currently deployed on `devel`*.
+
+When want to integrate to a newer contract, we'll:
+- make a PR to `elixir-omg/master` bumping the contract version in `mix.lock` and (possibly) providing compatibility with any possible breaking changes.
+This will run CI checks on the new integration
+- PR is merged
+- this is the moment when `elixir-omg/master` might not sync to `devel` anymore, if changes to contract were breaking!
+If they weren't breaking, we can end here, all should sync fine
+- if changes were breaking:
+- fast forward `plasma-contracts/v0.2` to the desired commit on `master`
```

Comment:
Can we use the same branch pairs here? If this for a PR in `elixir-omg/master`, then wouldn't it be pointing to a fixed SHA on `plasma-contracts`? Not necessarily a release branch (`v0.2`).

Reply:
I'm not sure I know what you mean here.

Let me rephrase perhaps - for the sake of explaining myself and also to get a better grip of the process (it's quite dizzying...)

I think the difference between `v0.2` and `master` in this example, is that `master` moves the quickest in both repos.

In contracts - `master` is always a set of building/test-passing code with the latest features/updates/fixes.

In elixir-omg - `master` - same applies to ch-ch/watcher. Now the trick is, that `elixir-omg/master` doesn't always use the topmost `plasma-contracts/master` SHA - because of the `mix.lock` and to have reproducible CI etc etc. So to move that forward with this, there must be this explicit PR to `elixir-omg/master`.

Now when this PR passes CI, we know that _should the `development` contract be redeployed from scratch_ on merge, everything should work fine. But we are not sure whether this change is backwards compatible, i.e. the new `elixir-omg/master` will work with the currently deployed `development`.

If it is expected to work, then nothing needs to happen - we can just continue to use the "old" `development` with the new `master`.

But if either it isn't or we want to test the new `v0.2` on a deployed env, we must redeploy the `development`. This is where `plasma-contracts/v0.2` would "come into play" - it sits at the spot, which is the SHA that is currently deployed on `development`. And now _should we decide to feature freeze the contract here_ this is the spot where `v0.2` would remain forever, ultimately being used to deploy `staging`/`Ari` (maybe we could allow to have well-reviewed hot-fixes sneak in there, but not sure).

I've rephrased the last step of the process a bit to maybe be clearer.

Reply:
> If it is expected to work, then nothing needs to happen - we can just continue to use the "old" development with the new master.

This is not a state we want to be in. It's going to cause confusion.

If `mix.lock` is updated with a new SHA on `plasma-contracts` repo, want to enforce that the _development_ environment *always* matches the `elixir-omg/master`, including its dependencies. This may require manual deployment steps now, but the goal is to automate this.

> But if either it isn't or we want to test the new v0.2 on a deployed env, we must redeploy the development.

There is an assumption here that is making things confusing. If we need to test `elixir-omg/v0.2` on a deployed env, we won't use the _development_ environment. We would provision a new environment to test, as necessary.

Fewer variables for these environments means less confusion. We want to declare that _development_ deploys `elixir-omg/master` and its dependencies, always.